content/ADR-204-comms-architecture.md

```diff
@@ -0,0 +1,620 @@
+
+        public Vector2Int Decode(int index) =>
+            new ((index % width) + MinX, (index / width) + MinY);
+```
+
+`terrainData.borderPadding = 2` in the new client.
+
+**Velocity Compression**
+
+```csharp
+        private static int CompressedVelocity(float velocity, int range, int sizeInBits)
+        {
+            int withoutSignBits = sizeInBits - 1;
+            float absVelocity = Mathf.Abs(velocity);
+            int compressed = FloatQuantizer.Compress(absVelocity, 0, range, withoutSignBits);
+            compressed <<= 1;
+            compressed |= NegativeSignFlag(velocity);
+            return compressed;
+        }
+
+        private static float DecompressedVelocity(int compressed, int range, int sizeInBits)
+        {
+            bool negativeSign = (compressed & 1) == 1;
+            int withoutSign = compressed >> 1;
+            int withoutSignBits = sizeInBits - 1;
+            float decompressed = FloatQuantizer.Decompress(withoutSign, 0, range, withoutSignBits);
+            if (negativeSign) decompressed *= -1;
+            return decompressed;
+        }
+```
+
+`CompressedVelocity`:
+- Takes a float velocity, a range, and bit size as inputs.
+- Calculates the absolute value of the velocity and quantizes it into an integer representation using a specified range and bit precision (excluding the sign bit).
+- Encodes the sign of the original velocity into the least significant bit (LSB).
+- Returns a compressed integer representing both the magnitude and sign.
+
+`DecompressedVelocity`:
+- Takes a compressed integer, a range, and bit size as inputs.
+- Extracts the sign information from the LSB and shifts the remaining bits to isolate the magnitude.
+- Decompresses the magnitude back to a float using the original range and precision.
+- Restores the original sign of the velocity and returns the decompressed value.
+
+These functions achieve efficient storage and retrieval of signed floating-point velocities by encoding the sign in the LSB and compressing the magnitude with quantization.
+
+### Movement
+
+```protobuf
+message Movement {
+  // command number
+  float timestamp = 1;
+  // world position
+  float position_x = 2;
+  float position_y = 3;
+  float position_z = 4;
+  // velocity
+  float velocity_x = 5;
+  float velocity_y = 6;
+  float velocity_z = 7;
+  // animations
+  float movement_blend_value = 8;
+  float slide_blend_value = 9;
+  bool is_grounded = 10;
+  bool is_jumping = 11;
+  bool is_long_jump = 12;
+  bool is_long_fall = 13;
+  bool is_falling = 14;
+
+  bool is_stunned = 15;
+  
+  float rotation_y = 16;
+}
+```
+
+Uncompressed `Movement` message can be used for debugging purposes, and it is optional to implement.
+It represents right the same values as the compressed version but without quantization and other compression techniques.
+
+Implementing this message is not sufficient to be compatible with the new client and its processing can be fully disabled in the production mode.
```

Comment:
If we want to keep this quite generic and abstract, we probably should not mention new client here, and rather mention that MovementCompressed must be implemented. Maybe something like that?
```suggestion
Implementing this message is not sufficient to be compatible with this version of comms as its processing can be fully disabled in the production mode. To ensure full functionality we should implement `MovementCompressed` as a main way to pass the data between clients
```